src/index.ts

```diff
@@ -0,0 +1,70 @@
+import React from 'react'
+import shallowEqual from './shallowEqual'
+
+type StateListener<T> = (state: T) => void
+type StateSelector<T, U> = (state: T) => U
+type PartialState<T> = Partial<T> | ((state: T) => Partial<T>)
+
+export default function create<
+  State extends Record<string, any>,
+  SetState extends (partialState: PartialState<Record<string, any>>) => void,
+  GetState extends () => Record<string, any>
+>(createState: (set: SetState, get: GetState) => State) {
+  const listeners: Set<StateListener<State>> = new Set()
+
+  const setState = (partialState: PartialState<State>) => {
+    state = Object.assign(
+      {},
+      state,
+      typeof partialState === 'function' ? partialState(state) : partialState
+    )
+    listeners.forEach(listener => listener(state))
+  }
+
+  const getState = () => state
+
+  const subscribe = (listener: StateListener<State>) => {
+    listeners.add(listener)
+    return () => {
+      listeners.delete(listener)
+    }
+  }
+
+  const destroy = () => {
+    listeners.clear()
+    state = {} as State
+  }
+
+  function useStore(): State
+  function useStore<U>(selector: StateSelector<State, U>): U
+  function useStore<U>(selector?: StateSelector<State, U>) {
+    // Gets entire state if no selector was passed in
+    const selectState = typeof selector === 'function' ? selector : getState
+    // Using functional initial b/c selected itself could be a function
+    const [stateSlice, setStateSlice] = React.useState(() => selectState(state))
+    // Prevent useEffect from needing to run when values change by storing them in a ref object
+    const refs = React.useRef({ stateSlice, selectState }).current
+
+    if (refs.stateSlice !== stateSlice) refs.stateSlice = stateSlice
+    if (refs.selectState !== selectState) refs.selectState = selectState
+
+    React.useEffect(() => {
+      return subscribe(() => {
```

Comment:
Is it supposed to receive the updated `state` in this listener?

Reply:
`state` is a `let` variable so it can be reassigned. It is declared once on `create()`. The callback passed into `subscribe` will always have the latest `state`.

Reply:
So, not arguing just trying to learn...

The useEffect will only be run once because the dependencies is an empty array. That means subscribe will only be called once, which means the listener function will only be created once. Won't it lock the state variable to what it was when the listener was created using JavaScript scoping? 10+ doing JavaScript and I'm still learning about scoping...

Reply:
No problem, I'm happy to share :)
This listener is created once but `let` and `var` variables declared in a parent scope are not frozen to their values. MDN has a good explanation of this:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures#Emulating_private_methods_with_closures